### PR TITLE
 Make CSS cacheable by browsers and proxies

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Make CSS cacheable by browsers and proxies (unless in debug mode).
This requires to have a response that does not depend on session (i.e. display settings), so I had to split main styles from palette and highcontrast.

### CODE REVIEW

File: inc/html.class.php [hunk 1]
```diff
@@ -6688,74 +6692,57 @@ public static function compileScss($args) {
...
-         $theme = 'auror';
-         if (isset($_SESSION["glpipalette"])) {
-            $theme = $_SESSION['glpipalette'];
-         }
-         $ckey .= '_' . $theme;
-         $files[] = 'css/palettes/' . $theme;
-      } else {
-         $ckey .= '_' . md5($args['file']);
+      $ckey .= '_' . md5($file);
+      $ckey = 'css_' . md5($ckey);
```
Comment (orthagh) [inline inc/html.class.php]: This double md5 seems to be overkill, maybe one only is sufficient now

File: inc/html.class.php [hunk 2]
```diff
@@ -6688,74 +6692,57 @@ public static function compileScss($args) {
...
+         $file .= '.scss';
+      }
 
-         if (!Toolbox::endsWith($args['file'], '.scss')) {
-            // Prevent include of file if ext is not .scss
-            $args['file'] .= '.scss';
-         }
+      // Requested file path
+      $path = GLPI_ROOT . '/' . $file;
+      $path = realpath($path);
```
Comment (orthagh) [inline inc/html.class.php]: relatpath added ?

File: inc/html.class.php [hunk 3]
```diff
@@ -6688,74 +6692,57 @@ public static function compileScss($args) {
...
+
+      // Alternate file path (prefixed by a "_", i.e. "_highcontrast.scss").
+      $pathargs = explode('/', $file);
+      $pathargs[] = '_' . array_pop($pathargs);
+      $pathalt = GLPI_ROOT . '/' . implode('/', $pathargs);
+      $pathalt = realpath($pathalt);
```
Comment (orthagh) [inline inc/html.class.php]: realtpath added?

File: inc/html.class.php (hunk 1 above)
Comment (cedric-anne) [reply inc/html.class.php]: Fixed.

File: inc/html.class.php (hunk 2 above)
Comment (cedric-anne) [reply inc/html.class.php]: I moved it [here](https://github.com/glpi-project/glpi/pull/5872/files#diff-298092ebe134c5b9ccbc7042328fb02cR6722). It is used to check that requested file is inside GLPI tree.

**Comment by orthagh** [comment]
switch from normal mode to debug mode doesn't remove browser cache

**Comment by orthagh** [comment]
On the same idea, user's picture are not cached currently, and on tickets list, i think it takes 50% more time to reload them (on my side 2s vs 1s).

I think we can cache (in browser) only theses pictures files if possible, not all document sending

**Comment by cedric-anne** [comment]
> switch from normal mode to debug mode doesn't remove browser cache

I fixed it by forcing presence of `debug` parameter in URL while in debug mode (see commit a860e632913cf4c72216500b2011a93d5d4ebb1b).

**Comment by trasher** [comment]
Broken on master.

**Pull request merged**